src/txn/txn_timestamp.c

```diff
@@ -198,24 +198,29 @@ __wt_txn_parse_timestamp(WT_SESSION_IMPL *session, const char *name,
  */
 static int
 __txn_get_pinned_timestamp(
-   WT_SESSION_IMPL *session, wt_timestamp_t *tsp, bool include_checkpoint)
+   WT_SESSION_IMPL *session, wt_timestamp_t *tsp, bool include_checkpoint,
+   bool include_oldest)
 {
 	WT_CONNECTION_IMPL *conn;
 	WT_TXN *txn;
 	WT_TXN_GLOBAL *txn_global;
 
 	conn = S2C(session);
 	txn_global = &conn->txn_global;
+	__wt_timestamp_set_zero(tsp);
 
-	if (!txn_global->has_oldest_timestamp)
+	if (include_oldest && !txn_global->has_oldest_timestamp)
 		return (WT_NOTFOUND);
+
 	__wt_readlock(session, &txn_global->rwlock);
-	__wt_timestamp_set(tsp, &txn_global->oldest_timestamp);
+	if (include_oldest)
+		__wt_timestamp_set(tsp, &txn_global->oldest_timestamp);
 
 	/* Check for a running checkpoint */
 	if (include_checkpoint &&
 	    !__wt_timestamp_iszero(&txn_global->checkpoint_timestamp) &&
-	    __wt_timestamp_cmp(&txn_global->checkpoint_timestamp, tsp) < 0)
+	    (!include_oldest ||
```

Comment:
I think better not to use include_oldest in the checkpoint related block, it will make coupling between whether to consider oldest timestamp or checkpoint timestamp. (i.e. instead of include_oldest, might be __wt_timestamp_iszero(tsp) will suffice.